File: ska_frontend/src/gui/page_assistant/dtos/dto_chat_details.ts
import { UtilsTypes } from "@/utils/core/utils_types";

export class DtoChatDetails {
    chat_id: number | null;
    user_id: number;
    chat_name: string;
    llm_model: string;
    prompt_template: string | null;
    temperature: number | null;
    top_p: number | null;
    default_prompt: string;

    constructor(obj: {
        chat_id: number | null,
        user_id: number,
        chat_name: string,
        llm_model: string,
        prompt_template: string | null,
        temperature: number | null,
        top_p: number | null,
        default_prompt: string,
    }) {
        this.chat_id = obj.chat_id;
        this.user_id = obj.user_id;
        this.chat_name = obj.chat_name;
        this.llm_model = obj.llm_model;
        this.prompt_template = obj.prompt_template;
        this.temperature = obj.temperature;
        this.top_p = obj.top_p;
        this.default_prompt = obj.default_prompt;
    }

    static fromUnknown(value: unknown): DtoChatDetails {
        const obj = UtilsTypes.unknownToObject(value).unwrap();
        return new DtoChatDetails({
            chat_id: UtilsTypes.unknownToNumberNullable(obj.chat_id).unwrap(),
            user_id: UtilsTypes.unknownToNumber(obj.user_id).unwrap(),
            chat_name: UtilsTypes.unknownToString(obj.chat_name).unwrap(),
            llm_model: UtilsTypes.unknownToString(obj.llm_model).unwrap(),
            prompt_template: UtilsTypes.unknownToStringNullable(obj.prompt_template).unwrap(),
            temperature: UtilsTypes.unknownToNumberNullable(obj.temperature).unwrap(),
            top_p: UtilsTypes.unknownToNumberNullable(obj.top_p).unwrap(),
            default_prompt: UtilsTypes.unknownToString(obj.default_prompt).unwrap(),
        })
    }
}


export class DtoCreateUpdateChatResponse {
    chat_id: number;

    constructor(obj: {
        chat_id: number,
    }) {
        this.chat_id = obj.chat_id;
    }

    static fromUnknown(value: unknown): DtoCreateUpdateChatResponse {
        const obj = UtilsTypes.unknownToObject(value).unwrap();
        return new DtoCreateUpdateChatResponse({
            chat_id: UtilsTypes.unknownToNumber(obj.chat_id).unwrap()
        })
    }
}

File: ska_backend/src/domain/user/dto_chat_details.rs
use std::str::FromStr;

use serde::{Deserialize, Serialize};

use crate::domain::{
    entities::{chat_message, user_chat},
    nn_model::{service_nn_model, DocumentDto},
};

use super::user_enums::ChatPacketType;

#[derive(Clone, Serialize, Deserialize)]
pub struct DtoChatDetails {
    pub chat_id: Option<i64>,
    pub user_id: i64,
    pub chat_name: String,
    pub llm_model: String,
    pub prompt_template: Option<String>,
    pub temperature: Option<f64>,
    pub top_p: Option<f64>,
    pub default_prompt: String,
}

impl DtoChatDetails {
    pub fn from_user_chat(chat: &user_chat::Model) -> DtoChatDetails {
        let llm_models = service_nn_model::get_nn_models_list();
        let llm_model = llm_models
            .into_iter()
            .find(|llm_model| llm_model.name == chat.llm_model);
        let default_prompt = match llm_model {
            Some(llm_model) => llm_model.default_prompt,
            None => "".to_string(),
        };
        DtoChatDetails {
            chat_id: Some(chat.chat_id),
            user_id: chat.user_id_fk,
            chat_name: chat.chat_name.clone(),
            llm_model: chat.llm_model.clone(),
            prompt_template: chat.prompt.clone(),
            temperature: chat.temperature,
            top_p: chat.top_p,
            default_prompt: default_prompt,
        }
    }
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DtoCreateUpdateChatResponse {
    pub chat_id: i64,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DtoChatPacket {
    pub created_at: i64,
    pub message_body: String,
    pub packet_type: ChatPacketType,
    pub context: Vec<DocumentDto>,
}

impl DtoChatPacket {
    pub fn from_chat_message(message: &chat_message::Model) -> anyhow::Result<DtoChatPacket> {
        let packet_type = ChatPacketType::from_str(message.message_type.as_str())
            .unwrap_or(ChatPacketType::Answer);

        let context: Vec<DocumentDto> = serde_json::from_value(message.context.clone())?;
        return Ok(DtoChatPacket {
            created_at: message.created_at.and_utc().timestamp(),
            message_body: message.message_body.clone(),
            packet_type: packet_type,
            context: context,
        });
    }
}
